File: frontend/src/types/order.ts
export type OrderStatus =
  | 'PENDING'
  | 'CONFIRMED'
  | 'PROCESSING'
  | 'SHIPPED'
  | 'DELIVERED'
  | 'CANCELLED'

export interface OrderItem {
  productId: string
  productName: string
  quantity: number
  unitPrice: number
  subtotal: number
}

export interface ShippingAddress {
  street: string
  city: string
  province: string
  postalCode: string
  country: string
}

export interface Order {
  id: string
  customerId: string
  customerName: string
  customerEmail: string
  items: OrderItem[]
  status: OrderStatus
  totalAmount: number
  shippingAddress: ShippingAddress
  notes?: string
  createdAt: string
  updatedAt: string
}

export interface OrderStats {
  pending: number
  confirmed: number
  processing: number
  shipped: number
  delivered: number
  cancelled: number
  total: number
}

export interface CreateOrderPayload {
  customerId: string
  customerName: string
  customerEmail: string
  items: {
    productId: string
    productName: string
    quantity: number
    unitPrice: number
  }[]
  shippingAddress: ShippingAddress
  notes?: string
}

export interface PageResponse<T> {
  content: T[]
  page: number
  size: number
  totalElements: number
  totalPages: number
  last: boolean
}

export interface ApiResponse<T> {
  success: boolean
  message?: string
  data: T
  timestamp: string
}

// WebSocket event types
export type WsEventType = 'ORDER_CREATED' | 'ORDER_STATUS_UPDATED' | 'ORDER_CANCELLED' | 'SUBSCRIBED'

export interface WsMessage {
  type: WsEventType
  orderId: string
  data: unknown
}

File: backend/src/main/java/com/orderflow/model/Order.java
package com.orderflow.model;

import lombok.*;
import org.springframework.data.annotation.*;
import org.springframework.data.mongodb.core.mapping.Document;
import org.springframework.data.mongodb.core.mapping.Field;
import org.springframework.data.mongodb.core.index.Indexed;

import java.math.BigDecimal;
import java.time.LocalDateTime;
import java.util.List;

@Document(collection = "orders")
@Data
@Builder
@NoArgsConstructor
@AllArgsConstructor
public class Order {

    @Id
    private String id;

    @Indexed
    private String customerId;

    private String customerName;
    private String customerEmail;

    private List<OrderItem> items;

    @Indexed
    private OrderStatus status;

    private BigDecimal totalAmount;

    private ShippingAddress shippingAddress;

    private String notes;

    // Optimistic locking — prevents concurrent update conflicts
    @Version
    private Long version;

    @CreatedDate
    private LocalDateTime createdAt;

    @LastModifiedDate
    private LocalDateTime updatedAt;

    // ─── Nested types ────────────────────────────────────────────────────────

    @Data
    @Builder
    @NoArgsConstructor
    @AllArgsConstructor
    public static class OrderItem {
        private String productId;
        private String productName;
        private Integer quantity;
        private BigDecimal unitPrice;

        public BigDecimal getSubtotal() {
            return unitPrice.multiply(BigDecimal.valueOf(quantity));
        }
    }

    @Data
    @Builder
    @NoArgsConstructor
    @AllArgsConstructor
    public static class ShippingAddress {
        private String street;
        private String city;
        private String province;
        private String postalCode;
        private String country;
    }

    // ─── Domain logic ────────────────────────────────────────────────────────

    public boolean canTransitionTo(OrderStatus newStatus) {
        return switch (this.status) {
            case PENDING    -> newStatus == OrderStatus.CONFIRMED || newStatus == OrderStatus.CANCELLED;
            case CONFIRMED  -> newStatus == OrderStatus.PROCESSING || newStatus == OrderStatus.CANCELLED;
            case PROCESSING -> newStatus == OrderStatus.SHIPPED;
            case SHIPPED    -> newStatus == OrderStatus.DELIVERED;
            default         -> false; // DELIVERED, CANCELLED are terminal
        };
    }

    public BigDecimal calculateTotal() {
        if (items == null) return BigDecimal.ZERO;
        return items.stream()
                .map(OrderItem::getSubtotal)
                .reduce(BigDecimal.ZERO, BigDecimal::add);
    }
}